<a href="https://colab.research.google.com/github/kangdoung/CS114.K21/blob/master/Sarcasm_LogisticRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nhóm :
* Dương Lê Tường Khang - 18520882
* Bùi Đào Gia Huy - 18520818
* Lã Trường Hải - 18520698

In [17]:
from google.colab import files
files.upload()

Saving kaggle.json to kaggle (1).json


{'kaggle.json': b'{"username":"khangdlt","key":"3dc1f86a2ffc9a36ce11a32beed2db24"}'}

Upload file chứa key của kaggle

In [2]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

Cài đặt môi trường

In [3]:
!kaggle datasets download -d rmisra/news-headlines-dataset-for-sarcasm-detection


  0% 0.00/3.30M [00:00<?, ?B/s]
100% 3.30M/3.30M [00:00<00:00, 101MB/s]


Dowload dataset từ kaggle về.

In [4]:
!unzip news-headlines-dataset-for-sarcasm-detection.zip

Archive:  news-headlines-dataset-for-sarcasm-detection.zip
  inflating: Sarcasm_Headlines_Dataset.json  
  inflating: Sarcasm_Headlines_Dataset_v2.json  


Giải nén dataset thành 2 file 
* Sarcasm_Headlines_Dataset.json 
* Sarcasm_Headlines_Dataset_v2.json

---



# Dataset của bài toán.

## Mô tả dataset


### **1. Mô tả bài toán**:
&ensp;&ensp;Phân loại 1 tựa đề của bài báo là *châm biếm (sarcastic)* hay *không châm biếm (non-sarcastic)*.
### **2. Cách thu thập dữ liệu train**:
Tiêu đề các bài báo được thu thập từ 2 trang:
* HuffPost: các tiêu đề được thu thập từ trang báo này sẽ được dán nhãn 0 ( non-sarcastic).
* TheOnion: nhãn 1 (sarcastic) sẽ được dán cho những tiêu đề được thu thập.Nhãn được thu thập ở 2 mục Brief và Photos Categories


## Đọc dataset

In [5]:
import pandas as pd
f1=pd.read_json('Sarcasm_Headlines_Dataset.json',lines=True)
f2=pd.read_json('Sarcasm_Headlines_Dataset_v2.json',lines=True)

Đọc file json từ thư viện pandas

In [6]:
f1.head()

article_link  ... is_sarcastic
0  https://www.huffingtonpost.com/entry/versace-b...  ...            0
1  https://www.huffingtonpost.com/entry/roseanne-...  ...            0
2  https://local.theonion.com/mom-starting-to-fea...  ...            1
3  https://politics.theonion.com/boehner-just-wan...  ...            1
4  https://www.huffingtonpost.com/entry/jk-rowlin...  ...            0

[5 rows x 3 columns]

In [7]:
f2.head()

is_sarcastic  ...                                       article_link
0             1  ...  https://www.theonion.com/thirtysomething-scien...
1             0  ...  https://www.huffingtonpost.com/entry/donna-edw...
2             0  ...  https://www.huffingtonpost.com/entry/eat-your-...
3             1  ...  https://local.theonion.com/inclement-weather-p...
4             1  ...  https://www.theonion.com/mother-comes-pretty-c...

[5 rows x 3 columns]

In [ ]:
f2

is_sarcastic  ...                                       article_link
0                 1  ...  https://www.theonion.com/thirtysomething-scien...
1                 0  ...  https://www.huffingtonpost.com/entry/donna-edw...
2                 0  ...  https://www.huffingtonpost.com/entry/eat-your-...
3                 1  ...  https://local.theonion.com/inclement-weather-p...
4                 1  ...  https://www.theonion.com/mother-comes-pretty-c...
...             ...  ...                                                ...
28614             1  ...  https://www.theonion.com/jews-to-celebrate-ros...
28615             1  ...  https://local.theonion.com/internal-affairs-in...
28616             0  ...  https://www.huffingtonpost.com/entry/andrew-ah...
28617             1  ...  https://www.theonion.com/mars-probe-destroyed-...
28618             1  ...  https://www.theonion.com/dad-clarifies-this-no...

[28619 rows x 3 columns]

In [8]:
df=pd.concat([f1,f2],axis=0,sort=False)
df.describe()

is_sarcastic
count  55328.000000
mean       0.458321
std        0.498264
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000

Nhóm 2 file Sarcasm_Headlines_Dataset.json và Sarcasm_Headlines_Dataset_v2.json chung với nhau

In [9]:
df=df.drop('article_link',axis=1)
df

headline  is_sarcastic
0      former versace store clerk sues over secret 'b...             0
1      the 'roseanne' revival catches up to our thorn...             0
2      mom starting to fear son's web series closest ...             1
3      boehner just wants wife to listen, not come up...             1
4      j.k. rowling wishes snape happy birthday in th...             0
...                                                  ...           ...
28614       jews to celebrate rosh hashasha or something             1
28615  internal affairs investigator disappointed con...             1
28616  the most beautiful acceptance speech this week...             0
28617  mars probe destroyed by orbiting spielberg-gat...             1
28618                 dad clarifies this not a food stop             1

[55328 rows x 2 columns]

Loại bỏ đi cột "article_link" do trong quá trình xử lý không cần đến.

In [10]:
X=df['headline']
y=df['is_sarcastic']

Lựa chọn:
* cột "headline" lựa chọn làm feature
* cột "is_sarcastic" lựa chọn làm label


In [11]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(df['headline'],df['is_sarcastic'],test_size=0.2,random_state=101)

Chia thành 2 tập test và train

In [12]:
from sklearn.feature_extraction.text import TfidfTransformer,CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

Cài đặt các thư viện cần thiết

In [13]:
pipe = Pipeline([('vect',CountVectorizer()),
               ('tfidf',TfidfTransformer()),
               ('model',LogisticRegression())])

model = pipe.fit(X_train, y_train)

Thay vì làm các công đoạn riêng, pipeline cho phép gộp các công đoạn lại với nhau.
* Counvectorizer: chuyển text thành vector
* TFidfTransformer: tính giá trị idf cho các vector trên
* LogisticRegression: model lựa chọn để train

Có thể thêm nhiều bước trung gian

Khi model gọi fit thì các object 'vect' và 'tfidf' sẽ tiến hành gọi 2 method fit và transform. Và ở bước cuối cùng thì object 'model' sẽ gọi method fit để tối ưu các tham số.

In [14]:
pred=model.predict(X_test)
print(accuracy_score(y_test,pred))

0.8942707392011567


Đánh giá của model dựa trên accuracy

Accuracy là độ đo được đánh giá bằng cách (số câu dự đoán đúng) / (toàn bộ câu của văn bản)


In [15]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,pred)

array([[5422,  567],
       [ 603, 4474]])

Ma trận trên cho ta thấy:
* Nhãn 0: dự đoán đúng 5422 hdlines và 567 hdlines bị dự đoán sai
* Nhãn 1: tương tự, có 603 hdlines bị dự đoán sai và 4474 hdlines được dự đoán đúng.

## Sarcasm detect  1 headline bất kì dựa trên model Logistic Regression

In [ ]:
import numpy as np

In [ ]:
hdlines = input()
hdline_pre = [hdlines]
flag = model.predict(hdline_pre)
if flag == 1:
  print("sarcastic")
else:
  print("non-sarcastic")

Civil Whites
is sarcastic


# Predict trên 2000 dữ liệu crawl về từ 2 trang web như tập dataset trên.

## Cách thu thập dữ liệu:
Xây dựng bộ dataset tương tự như tập Train bằng cách lấy tiêu đề của các bài báo tử 2 trang báo Huffpost and TheOninon.
### Crawl dữ liệu:
1. Tư 2 url của trang báo trích xuất file "html".
2. Từ file html lựa chọn các link và headline để gán nhãn theo quy tắc sau:
  2.1. Các tiêu đề từ bài báo Huffpost dán nhãn 0  
  2.2. Các tiêu đề từ bài báo TheOnion dán nhãn 1  
    
Link dataset:
[dataset](https://github.com/kangdoung/CS114.K21/tree/master/Crawl). 

## Dữ liệu thu thập


In [18]:
import pandas as pd
df1 = pd.read_csv("data_crawl.csv")
df1

Article  ... is_sarcasm
0     https://www.huffpost.com/entry/unemployment-nu...  ...          0
1     https://ogn.theonion.com/10-clumps-of-grass-in...  ...          1
2     https://ogn.theonion.com/10-clumps-of-grass-in...  ...          1
3     https://ogn.theonion.com/10-clumps-of-grass-in...  ...          1
4     https://ogn.theonion.com/10-clumps-of-grass-in...  ...          1
...                                                 ...  ...        ...
1995  https://local.theonion.com/zoom-crasher-become...  ...          1
1996  https://www.huffpost.com/entry/tik-tok-kpop-zo...  ...          0
1997  https://www.huffpost.com/entry/zuckerberg-revi...  ...          0
1998  https://www.huffpost.com/entry/zuckerberg-revi...  ...          0
1999  https://www.huffpost.com/entry/zuckerberg-defe...  ...          0

[2000 rows x 3 columns]

Dữ liệu mới thu thập

In [19]:
df1=df1.drop('Article',axis=1)


In [ ]:
df1

Headlines  is_sarcasm
0     15 Million More Americans Sought Unemployment ...           0
1                     10 Clumps Of Grass In Video Games           1
2                     10 Clumps Of Grass In Video Games           1
3                     10 Clumps Of Grass In Video Games           1
4                     10 Clumps Of Grass In Video Games           1
...                                                 ...         ...
1995  Zoom Crasher Becomes Too Engrossed In Sales Me...           1
1996  Zoomers Boast They Sabotaged Trump Rally Turno...           0
1997  Zuckerberg Says Hell Review Policies That Allo...           0
1998  Zuckerberg Says Hell Review Policies That Allo...           0
1999  Zuckerberg Says Trumps Inflammatory Facebook P...           0

[2000 rows x 2 columns]

Tương tự loại bỏ cột Link.

In [ ]:
df1.columns

Index(['Headlines', 'is_sarcasm'], dtype='object')

Các cột có trong dữ liệu

In [20]:
nX=df1['Headlines']
ny=df1['is_sarcasm']
ny

0       0
1       1
2       1
3       1
4       1
       ..
1995    1
1996    0
1997    0
1998    0
1999    0
Name: is_sarcasm, Length: 2000, dtype: int64

Tương tự chọn headline làm feature và chọn is_sarcasm làm label

# Logistic Regression

In [21]:
n_model = pipe.fit(X,y)

Train lại model trên với tập dữ liệu ban đầu

In [22]:
pred = n_model.predict(nX)

In [23]:
accuracy_score(ny, pred)

0.7135

Kết quả thấp so với kết quả của model ban đầu. Có thể thấp model khá tệ do có hơn 500 headline đã dự đoán sai trong tổng số 2000 headlines.


In [29]:
confusion_matrix(ny,pred)

array([[858, 142],
       [431, 569]])

Ta có thể thấy từ ma trận trên, model dự đoán khá kém các headlines có nhãn 1 (sarcastic) và dự đoán khá tốt các headlines non-sarcastic.

# Multinomial Naive Bayes

In [24]:
from sklearn.naive_bayes import MultinomialNB

In [25]:
pl = Pipeline([
    ('bow', CountVectorizer()),
    ('tfidf', TfidfTransformer()),  
    ('classifier', MultinomialNB()),  
])

In [26]:
model = pl.fit(X,y)

In [30]:
predict = model.predict(nX)

In [31]:
accuracy_score(ny, predict)

0.708

Kết quả cho ra kém hơn model Logistic Regression.

In [32]:
confusion_matrix(ny, predict)

array([[898, 102],
       [482, 518]])

Model MultinomialNB dự đoán tốt hơn ở non-sarcastic nhưng lại tệ hơn ở nhãn sarcastic.

# **Nhận xét**:

Model cho kết quả kém có thể do xảy ra overfitting.

**Nguyên nhân**:

Dữ liệu cho tập train quá lớn (xấp xỉ 56k headlines) trong khi tập test chỉ có 2k headlines, dẫn đến việc mô hình bị overfitting. Các tỉ lệ thường dùng trong việc chia các tập dữ liệu là 80:20 hoặc 70:30.

**Hướng giải quyết**:
có thể tăng số lượng headlines tập test lên hoặc chia bớt số lượng bên tập train qua tập test.

**Model**:
Trong 2 model trên thì model LogisticRegression cho accuracy là 0.7135 cao hơn so với model MultinomialNB (accuraacy = 0.708). Theo một số tài liệu nhóm tìm hiểu được, thì NB hội tụ nhanh hơn nhưng lỗi cao hơn so với logistic Regression. Do đó với các bộ dữ liệu lớn, Logistic Regression cho kết quả tốt hơn, và ngược lại NB tốt so với các bộ dữ liệu nhỏ.